Complex shapes docs (#587)
trixi-framework/TrixiParticles.jl: docs/src/preprocessing/preprocessing.md

# Sample Geometry

Generating the initial configuration of a simulation requires filling volumes (3D) or surfaces (2D) of complex geometries with particles.
The algorithm to sample a complex geometry should be robust and fast,

> since for large problems (high number of particles) or complex geometries (many geometry-faces),

Reviewer: ```suggestion
since for large problems (high numbers of particles) or complex geometries (many geometry faces),
```

> generating the initial configuration is not trivial and can be very expensive in terms of computational costs.

Reviewer: ```suggestion
generating the initial configuration is not trivial and can be very expensive in terms of computational cost.
```

We therefore use a [winding number](https://en.wikipedia.org/wiki/Winding_number) approach for an inside-outside segmentation of an object.
The winding number $w(\mathbf{p})$ is a signed integer-valued property of a point $\mathbf{p}$ and is defined as

$w(\mathbf{p}) = \frac{1}{2 \pi} \sum^n_{i=1} \Theta_i,$

where  $\Theta_i$ is the **signed** angle between vectors from two consecutive vertices to $\mathbf{p}$.
In 3D, we refer to the solid angle of an *oriented* triangle with respect to $\mathbf{p}$

We provide the following methods to calculate $w(\mathbf{p})$:
- Horman et al. (2001) evaluated the winding number combined with an even-odd rule and is only for 2D polygons (see [WindingNumberHorman](@ref)).
- Naive winding: Jacobson et al. (2013) generalized the winding number so that the algorithm can be applied for 2D and also for 3D geometries (see [WindingNumberJacobson](@ref)).

> - Hierarchical winding: Jacobson et al. (2013) also introduced a fast hierarchical evaluation of the winding number. For further information see description below.

Reviewer: ```suggestion
- Hierarchical winding: Jacobson et al. (2013) also introduced a fast hierarchical evaluation of the winding number. For further information see the description below.
```

> ## [Hierarchical Winding](@id hierarchical_winding)

Author: thoughts?

</figure>
```

That means, if we want to calculate the winding number only for a specific area of interest, e.g. the top right square of a circle, we can use the closure of this area as we can see in the following image.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/8e1c2a6d-3b28-4f75-991f-0d0ecc9f6db1" alt="circle"/>
</figure>
```

For a continuous problem, we then have an open surface $\mathcal{S}$ and an arbitrary closing surface $\bar{\mathcal{S}}$, where $\partial \bar{\mathcal{S}} = \partial \mathcal{S}$,
such that $\mathcal{B} = \bar{\mathcal{S}} \cup \mathcal{S}$ is some closed oriented surface.
If an arbitrary query point $\mathbf{p}$ is outside the convex hull of $\mathcal{B}$, we know that

$w_{\mathcal{S}}(\mathbf{p}) + w_{\bar{\mathcal{S}}}(\mathbf{p}) = w_{\mathcal{B}}(\mathbf{p}) = 0.$

This means

$w_{\mathcal{S}}(\mathbf{p}) = - w_{\bar{\mathcal{S}}}(\mathbf{p}) ,$

regardless of how $\bar{\mathcal{S}}$ is constructed.

### Bounding volume hierarchy

If we construct a bounding volume hierarchy $T$ and evaluate $w$ hierarchical,

> we can perform asymptotically better than the naive winding computing $w$ for the whole geometry as is shown by Jacobson et al. (2013).

Reviewer: ```suggestion
we can perform asymptotically better than the naive winding number computation of $w$ for the whole geometry as is shown by Jacobson et al. (2013).
```

The algorithm behind (Jacobson et al., Algorithm 2, p. 5) calls itself recursively, where the recursion stops with the following criteria.

- if $T$ is a leaf then evaluate the winding number naively $w(\mathbf{p}, T.\mathcal{S})$ where $T.\mathcal{S}$ is the open surface in $T$.
- else if $\mathbf{p}$ is outside $T$ then evaluate the winding number naively $-w(\mathbf{p}, T.\bar{\mathcal{S}})$ where $T.\bar{\mathcal{S}}$ is the closing surface of $T$.

#### Continuous geometry

For an arbitrary continuous geometry, the steps are as follows.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/66c423ae-28ce-4d48-aa52-777726a3a677" alt="continuous closing"/>
</figure>
```

(1):
- Recurse left: $w_{\text{left}} = \text{\texttt{hierarchical\_winding}} (\mathbf{p}, T.\text{left})$
- Recurse right: $w_{\text{right}} = \text{\texttt{hierarchical\_winding}} (\mathbf{p},T.\text{right})$

(2):

> - Query point $\mathbf{p}$ is outside bounding box $T$.
> - Use closure $T.\bar{\mathcal{S}}$.
> - Note the negative winding $\rightarrow w_{\mathcal{S}}(\mathbf{p}) = - w_{\bar{\mathcal{S}}}(\mathbf{p})$
> 
> $w_{\text{left}} = -\text{\texttt{naive\_winding}} (\mathbf{p}, T.\bar{\mathcal{S}})$

Reviewer 2: ```suggestion
- Query point $\mathbf{p}$ is outside bounding box $T$, so don't recurse deeper.
- Compute $w_{\mathcal{S}}(\mathbf{p}) = - w_{\bar{\mathcal{S}}}(\mathbf{p})$ with the closure $T.\bar{\mathcal{S}}$, which is generally much smaller (fewer edges in the discrete version) than $T.\mathcal{S}$

$w_{\text{left}} = -\text{\texttt{naive\_winding}} (\mathbf{p}, T.\bar{\mathcal{S}})$
```

(3):

> - Bounding box $T$ is a leaf. Use open surface $T.\mathcal{S}$
> 
> $w_{\text{right}} = \text{\texttt{naive\_winding}} (\mathbf{p}, T.\mathcal{S})$

Reviewer 2: ```suggestion
- Bounding box $T$ is a leaf. Use open surface $T.\mathcal{S}$.

$w_{\text{right}} = \text{\texttt{naive\_winding}} (\mathbf{p}, T.\mathcal{S})$.
```
Please also add periods above to make these full sentences.

The reconstructed surface will then look as in the following image.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/920bb4f1-1336-4e77-b06d-d5b46ca0d8d5" alt="reconstructed surface"/>
</figure>
```

We finally sum up the winding numbers

$w = w_{\text{left}} + w_{\text{right} } = -w_{T_{\text{left}}.\bar{\mathcal{S}}} + w_{T_{\text{right}}.\mathcal{S}}$

> #### Discrete geometry

Reviewer 2: ```suggestion
#### Discrete example

We will now go through the discrete version of the example above.
```

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/a9b59cc3-5421-40af-b0b0-f4c18a5a7078" alt="discrete geometry"/>
</figure>
```

> To construct the hierarchy for discrete piecewise-linear surfaces (1), we have to do the following.

Reviewer 2: ```suggestion
To construct the hierarchy for the discrete piecewise-linear example in (1), we have to do the following.
```

(2):

> Each edge is distributed to the child whose box contains the edge's barycenter.

Reviewer 2: ```suggestion
Each edge is distributed to the child whose box contains the edge's barycenter (red dots in (2)).
```

Splitting stops when the number of a box's edges slips below a
threshold (usually $\approx 100$ faces in 3D, here: 6 edges).

(3):
For the closure, Jacobson et al. (2013) define *exterior vertices* (*exterior edges* in 3D)

> as boundary vertices of such a segmentation.

Reviewer 2: ```suggestion
as boundary vertices of such a segmentation (red dots in (3)).
```

To find them, we traverse around each edge (face in 3D) in order, and
increment or decrement for each vertex (edge) a specific counter.

> ```julia
>   # 2D
>   v1 = edge_vertices_ids[edge][1]
>   v2 = edge_vertices_ids[edge][2]
> 
>   vertex_count[v1] += 1
>   vertex_count[v2] -= 1
> ```
> ```julia
>   # 3D
>   edge1 = face_edges_ids[face][1]
> 
>   if edge_vertices_ids[edge1] == (v1, v2)
>       directed_edges[edge1] += 1
>   else
>       directed_edges[edge1] -= 1
>   end
> ```

Reviewer 2: ````suggestion
```julia
v1 = edge_vertices_ids[edge][1]
v2 = edge_vertices_ids[edge][2]

vertex_count[v1] += 1
vertex_count[v2] -= 1
```
````
1. Why the indent?
2. Do you want to show the 3D example here? I think it might be more confusing than helpful.

> In 2D, a vertex is declared as exterior if `vertex_count(vertex) != 0`.
> The exterior vertices are then connected to one arbitrary
> exterior vertex using appropriately oriented line segments
> 
> ```julia
>   edge = vertex_count[v] > 0 ? (closing_vertex, v) : (v, closing_vertex)
> ```
> 
> In 3D, an edge is declared as exterior if `directed_edges[edge] != 0`
> The exterior edges are triangulated with one exterior vertex $k$ with orientation
> - ``\{i, j, k\}`` if `directed_edges[edge] < 0`
> - ``\{j, i, k\}`` if `directed_edges[edge] > 0`

Reviewer 2: ````suggestion
In 2D, a vertex is declared as exterior if `vertex_count(vertex) != 0`, so there is not the same amount of edges in this box going into vs out of the vertex.
To construct the closing surface, the exterior vertices are then connected to one arbitrary
exterior vertex using appropriately oriented line segments

```julia
edge = vertex_count[v] > 0 ? (closing_vertex, v) : (v, closing_vertex)
```
The resulting closed surface ``T.S \cup T.\bar{S}`` then has the same number of edges going into and out of each vertex.
````
Again, I wouldn't show the 3D case.

#### Incorrect evaluation

If we follow the algorithm, we know that recursion stops if

> - The bounding box $T$ is a leaf
> - The query point $\mathbf{p}$ is outside the box

Reviewer 2: ```suggestion
- the bounding box $T$ is a leaf or
- the query point $\mathbf{p}$ is outside the box.
```
Either lowercase and comma/or or full sentences with a full stop.

> (1): If $\mathbf{p}$ is outside the box, we calculate the winding number with the (red) closure of the box.
> 
> (2): If $\mathbf{p}$ is inside the box, we use the (blue) edges distributed to the box.
> 
> (3): In this case, it leads to an incorrect evaluation of the winding number.

Reviewer 2: ```suggestion
(1): The query point $\mathbf{p}$ is outside the box, so we calculate the winding number with the (red) closure of the box.

(2): The query point $\mathbf{p}$ is inside the box, so we use the (blue) edges distributed to the box.

(3): In this case, it leads to an incorrect evaluation of the winding number.
The query point is clearly inside the box, but not inside the reconstructed surface.
This is because the property $w_{\mathcal{S}}(\mathbf{p}) = - w_{\bar{\mathcal{S}}}(\mathbf{p})$ only holds when $\mathbf{p}$ is outside of $\mathcal{B}$, which is not the case here.
```

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/7bae164a-8d5b-4761-9d54-9abf99fca94a" alt="incorrect evaluation"/>
</figure>

> ```

Reviewer 2: Why is this now below the explanation? Earlier you always showed the pictures first before explaining them.

#### Correct evaluation

> Jacobson et al. (2013) therefore resize the bounding box, since it doesn't matter if the boxes overlap.

Reviewer 2: ```suggestion
Jacobson et al. (2013) therefore resize the bounding box to fully include the closing surface of the neighboring box, since it doesn't matter if the boxes overlap.
```

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/097f01f4-1f37-48e4-968a-4c0970548b24" alt="correct evaluation resizing"/>
</figure>
```

> To avoid resizing, we calculate the closure of the bounding box differently:

Reviewer 2: ```suggestion
To avoid resizing, we take a different approach and calculate the closure of the bounding box differently:
```

$w_{\text{left}} = -\text{\texttt{naive\_winding}} (\mathbf{p}, T.\overline{\mathcal{S}})$

(3):
- Bounding box $T$ is a leaf. Use open surface $T.\mathcal{S}$:

$w_{\text{right}} = \text{\texttt{naive\_winding}} (\mathbf{p}, T.\mathcal{S})$.

The reconstructed surface will then look as in the following image.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/920bb4f1-1336-4e77-b06d-d5b46ca0d8d5" alt="reconstructed surface"/>
</figure>
```

We finally sum up the winding numbers

$w = w_{\text{left}} + w_{\text{right} } = -w_{T_{\text{left}}.\overline{\mathcal{S}}} + w_{T_{\text{right}}.\mathcal{S}}$

#### Discrete example

We will now go through the discrete version of the example above.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/a9b59cc3-5421-40af-b0b0-f4c18a5a7078" alt="discrete geometry"/>
</figure>
```

To construct the hierarchy for the discrete piecewise-linear example in (1), we have to do the following.

(2):

> Each edge is distributed to the child whose box contains the edge's overlineycenter (red dots in (2)).

Author: Yes, I thought it was fixed?

# Sampling of Geometries

Generating the initial configuration of a simulation requires filling volumes (3D) or surfaces (2D) of complex geometries with particles.
The algorithm to sample a complex geometry should be robust and fast,
since for large problems (high numbers of particles) or complex geometries (many geometry faces),
generating the initial configuration is not trivial and can be very expensive in terms of computational cost.
We therefore use a [winding number](https://en.wikipedia.org/wiki/Winding_number) approach for an inside-outside segmentation of an object.
The winding number $w(\mathbf{p})$ is a signed integer-valued function of a point $\mathbf{p}$ and is defined as

$w(\mathbf{p}) = \frac{1}{2 \pi} \sum^n_{i=1} \Theta_i,$

> where  $\Theta_i$ is the **signed** angle between vectors from two consecutive vertices $\mathbf{c}_i$ and $\mathbf{c}_{i+1}$ on a curve to $\mathbf{p}$.

Author: > where  ``\Theta_i`` is the **signed** angle between the directions of ``\mathbf{p}`` to two consecutive vertices ``\mathbf{c}_i`` and ``\mathbf{c}_{i+1}`` on a curve.

better?

> where  $\Theta_i$ is the **signed** angle between vectors from two consecutive vertices $\mathbf{c}_i$ and $\mathbf{c}_{i+1}$ on a curve to $\mathbf{p}$.

Reviewer 2: Do you mean the angle between the coordinate vectors of two consecutive vertices?

> where  $\Theta_i$ is the **signed** angle between vectors from two consecutive vertices $\mathbf{c}_i$ and $\mathbf{c}_{i+1}$ on a curve to $\mathbf{p}$.

Reviewer 2: Ah nvm, now I understood it.

> where  $\Theta_i$ is the **signed** angle between vectors from two consecutive vertices $\mathbf{c}_i$ and $\mathbf{c}_{i+1}$ on a curve to $\mathbf{p}$.

Reviewer 2: ```
where ``\Theta_i`` is the signed angle between ``\mathbf{c}_i - \mathbf{p}`` and ``\mathbf{c}_{i+1} - \mathbf{p}`` and ``\mathbf{c}_i`` and ``\mathbf{c}_{i+1}`` are two consecutive vertices on a curve.
```

> w(\mathbf{p}) = \frac{1}{2 \pi} \sum^n_{i=1} \Theta_i,

Reviewer 2: ```suggestion
w(\mathbf{p}) = \frac{1}{2 \pi} \sum^n_{i=1} \Theta_i.
```

```

> here, ``\Theta_i`` is the *signed* angle between ``\mathbf{c}_i - \mathbf{p}`` and ``\mathbf{c}_{i+1} - \mathbf{p}`` where ``\mathbf{c}_i`` and ``\mathbf{c}_{i+1}`` are two consecutive vertices on a curve.

Reviewer 2: ```suggestion
Here, ``\Theta_i`` is the *signed* angle between ``\mathbf{c}_i - \mathbf{p}`` and ``\mathbf{c}_{i+1} - \mathbf{p}`` where ``\mathbf{c}_i`` and ``\mathbf{c}_{i+1}`` are two consecutive vertices on a curve.
```

- Recurse right: $w_{\text{right}} = \text{\texttt{hierarchical\_winding}} (\mathbf{p},T.\text{right})$

(2):
- Query point $\mathbf{p}$ is outside bounding box $T$, so don't recurse deeper.
- Compute $w_{\mathcal{S}}(\mathbf{p}) = - w_{\overline{\mathcal{S}}}(\mathbf{p})$ with the closure $T.\overline{\mathcal{S}}$, which is generally much smaller (fewer edges in the discrete version) than $T.\mathcal{S}$:

$w_{\text{left}} = -\text{\texttt{naive\_winding}} (\mathbf{p}, T.\overline{\mathcal{S}})$

(3):
- Bounding box $T$ is a leaf. Use open surface $T.\mathcal{S}$:

> $w_{\text{right}} = \text{\texttt{naive\_winding}} (\mathbf{p}, T.\mathcal{S})$.

Reviewer 2: ```suggestion
$w_{\text{right}} = \text{\texttt{naive\_winding}} (\mathbf{p}, T.\mathcal{S})$
```
The period ends up in the next line for some reason.

Splitting stops when the number of a box's edges slips below a
threshold (usually $\approx 100$ faces in 3D, here: 6 edges).

(3):
For the closure, Jacobson et al. (2013) define *exterior vertices* (*exterior edges* in 3D)
as boundary vertices of such a segmentation (red dots in (3)).
To find them, we traverse around each edge (face in 3D) in order, and
increment or decrement for each vertex (edge) a specific counter.

```julia
v1 = edge_vertices_ids[edge][1]
v2 = edge_vertices_ids[edge][2]

vertex_count[v1] += 1
vertex_count[v2] -= 1
```

In 2D, a vertex is declared as exterior if `vertex_count(vertex) != 0`, so there is not the same amount of edges in this box going into versus out of the vertex.
To construct the closing surface, the exterior vertices are then connected to one arbitrary
exterior vertex using appropriately oriented line segments:

```julia
edge = vertex_count[v] > 0 ? (closing_vertex, v) : (v, closing_vertex)
```

The resulting closed surface ``T.S \cup T.\overline{S}`` then has the same number of edges going into and out of each vertex.

#### Incorrect evaluation

If we follow the algorithm, we know that recursion stops if

- the bounding box $T$ is a leaf or

> - the query point $\mathbf{p}$ is outside the box.

Reviewer 2: ```suggestion
- the query point ``\mathbf{p}`` is outside the box.
```
As mentioned before, please change all of these, as dollar signs are not officially supported.

```@raw html
<figure>
  <img src="https://github.com/user-attachments/assets/7bae164a-8d5b-4761-9d54-9abf99fca94a" alt="incorrect evaluation"/>
</figure>
```

(1): The query point $\mathbf{p}$ is outside the box, so we calculate the winding number with the (red) closure of the box.

(2): The query point $\mathbf{p}$ is inside the box, so we use the (blue) edges distributed to the box.

(3): In this case, it leads to an incorrect evaluation of the winding number.
The query point is clearly inside the box, but not inside the reconstructed surface.
This is because the property $w_{\mathcal{S}}(\mathbf{p}) = - w_{\overline{\mathcal{S}}}(\mathbf{p})$
only holds when $\mathbf{p}$ is outside of $\mathcal{B}$, which is not the case here.

#### Correct evaluation

> Jacobson et al. (2013) therefore resize the bounding box to fully include the closing surface
> of the neighboring box, since it doesn't matter if the boxes overlap.

Reviewer 2: ```suggestion
Jacobson et al. (2013) don't mention this problem or provide a solution to it.
We contacted the authors and found that they know about this problem and solve it by resizing the bounding box to fully include the closing surface
of the neighboring box, since it doesn't matter if the boxes overlap.
```
Otherwise, it sounds like they provide a solution in the paper.
